src/java/org/apache/cassandra/service/CleanupStateTracker.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class CleanupStateTracker
+{
+    private static final Logger log = LoggerFactory.getLogger(CleanupStateTracker.class);
+
+    @VisibleForTesting
+    static final String CLEANUP_STATE_FILE_NAME = "node_cleanup_state.json";
+    private static final Instant MIN_TS = Instant.parse("2022-01-01T00:00:00Z");
+
+    private final CleanupState cleanupStateCache;
+
+    public CleanupStateTracker() throws IOException
+    {
+        this.cleanupStateCache = new CleanupState(createOrRetrieveCleanupStateFile());
+    }
+
+    @VisibleForTesting
+    File createOrRetrieveCleanupStateFile() throws IOException
+    {
+        File cleanupStateFile = new File(cleanupStateFileLocation());
+        try
+        {
+            cleanupStateFile.createNewFile();
+            return cleanupStateFile;
+        }
+        catch (IOException e)
+        {
+            log.warn("Cannot retrieve or create node cleanup state file.", cleanupStateFile.getAbsolutePath(), e);
+            throw e;
+        }
+    }
+
+    @VisibleForTesting
+    String cleanupStateFileLocation(){
+        return DatabaseDescriptor.getPersistentSettingsLocation() + CLEANUP_STATE_FILE_NAME;
+    }
+
+    /** Creates table entry if it does not already exist */
+    public synchronized void createCleanupEntryForTableIfNotExists(String keyspaceName, String columnFamily)
```

Comment:
I think it's a bug for this and the next function to need to be synchronized.  Ideally you can contain the need for synchronization inside of the CleanupState class.